test/exclude-members.test.js

```diff
@@ -0,0 +1,163 @@
+    mfSeed: Buffer.from(
+      '000000000000000000000000000000000000000000000000000000000000a1ce',
+      'hex'
+    ),
+  })
+  const bob = Testbot({
+    keys: ssbKeys.generate(null, 'bob'),
+    mfSeed: Buffer.from(
+      '0000000000000000000000000000000000000000000000000000000000000b0b',
+      'hex'
+    ),
+  })
+
+  await alice.tribes2.start()
+  await bob.tribes2.start()
+  t.pass('tribes2 started for both alice and bob')
+
+  await p(alice.metafeeds.findOrCreate)()
+  const bobRoot = await p(bob.metafeeds.findOrCreate)()
+
+  await replicate(alice, bob)
+  t.pass('alice and bob replicate their trees')
+
+  const {
+    id: groupId,
+    writeKey: writeKey1,
+    subfeed: { id: firstFeedId },
+  } = await alice.tribes2.create().catch((err) => {
+    console.error('alice failed to create group', err)
+    t.fail(err)
+  })
+  t.pass('alice created a group')
+
+  await alice.tribes2.addMembers(groupId, [bobRoot.id]).catch((err) => {
+    console.error('add member fail', err)
+    t.fail(err)
+  })
+  t.pass('alice added bob to the group')
+
+  t.equals(
+    await p(countGroupFeeds)(alice),
+    1,
+    'before exclude alice has 1 group feed'
+  )
+
+  await alice.tribes2.excludeMembers(groupId, [bobRoot.id]).catch((err) => {
+    console.error('remove member fail', err)
+    t.fail(err)
+  })
+
+  t.equals(
+    await p(countGroupFeeds)(alice),
+    2,
+    'after exclude alice has 2 group feeds'
+  )
+
+  const {
+    value: { author: secondFeedId },
+  } = await alice.tribes2
+    .publish({
+      type: 'test',
+      text: 'post',
+      recps: [groupId],
+    })
+    .catch(t.fail)
+
+  t.equals(
+    await p(countGroupFeeds)(alice),
+    2,
+    'alice still has 2 group feeds after publishing on the new feed'
+  )
+
+  t.notEquals(
+    secondFeedId,
+    firstFeedId,
+    'feed for publish is different to initial feed'
+  )
+
+  const { writeKey: writeKey2 } = await alice.tribes2.get(groupId)
+
+  t.false(writeKey1.key.equals(writeKey2.key), "there's a new key for writing")
+
+  const msgsFromFirst = await alice.db.query(
+    where(author(firstFeedId)),
+    toPromise()
+  )
+
+  const firstContents = msgsFromFirst.map((msg) => msg.value.content)
+
+  t.equal(firstContents.length, 5, '5 messages on first feed')
+
+  const firstInit = firstContents[0]
+
+  t.equal(firstInit.type, 'group/init')
+  t.equal(firstInit.groupKey, writeKey1.key.toString('base64'))
+
+  //const addAlice = firstContents[1]
+  //TODO
+
+  //const addBob = firstContents[2]
+  //TODO
+
+  //const excludeMsg = firstContents[3]
+
+  // TODO: test excludeMsg once we use the correct format
+
+  //const reinviteMsg = firstContents[4]
+
+  // TODO: test reinviteMsg once we use the correct format (add-member)
+
+  const msgsFromSecond = await alice.db.query(
+    where(author(secondFeedId)),
+    toPromise()
+  )
+
+  const secondContents = msgsFromSecond.map((msg) => msg.value.content)
+  console.log({ firstContents, secondContents })
```

Comment:
This prints

```js
{
  firstContents: [
    {
      type: 'group/init',
      version: 'v2',
      groupKey: 'WmvadXKxC/g/ZDSL6fW6TXi+Edw6Yp/A2L0t6z6vZXc=',
      tangles: [Object]
    },
    {
      type: 'group/exclude',
      excludes: [Array],
      recps: [Array],
      tangles: [Object]
    }
  ],
  secondContents: [
    {
      type: 'group/init',
      version: 'v2',
      groupKey: 'j2Xhn9f1u63AebGOQvaa47A3M5seC+2BJJk41ka9A7I=',
      tangles: [Object],
      recps: [Array]
    },
    {
      type: 'group/move-epoch',
      secret: 'j2Xhn9f1u63AebGOQvaa47A3M5seC+2BJJk41ka9A7I=',
      exclusion: 'ssb:message/classic/fWSZ6QG9sX_sC9E7C6A-oBX1M_sbXuBGbW-UAaSGcFM=',
      recps: [Array],
      tangles: [Object]
    },
    { type: 'test', text: 'post', recps: [Array], tangles: [Object] }
  ]
}
```

The exclusion message should be on the first feed, that's a TODO. But other than that, where are the normal add-member messages? If nothing else, alice should've added herself to the group when creating it. There are even other tests that test that... Am I fetching the feed wrong or something?

Reply:
lmao forgot about the invitations feed